postgrenosql/src/main/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClient.java

```diff
@@ -0,0 +1,332 @@
+        connection.setAutoCommit(autoCommit);
+      } catch (Exception e) {
+        LOG.error("Error during initialization: " + e);
+      }
+    }
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    if (INIT_COUNT.decrementAndGet() == 0) {
+      postgrenosqlDriver = null;
+      LOG.info("Local cleaned up.");
+    }
+  }
+
+  @Override
+  public Status read(String tableName, String key, Set<String> fields, Map<String, ByteIterator> result) {
+    try{
+      PreparedStatement readStatement = connection.prepareStatement(createReadStatement(tableName, fields));
+      readStatement.setString(1, key);
+
+      ResultSet resultSet = readStatement.executeQuery();
+      if (!resultSet.next()) {
+        resultSet.close();
+        return  Status.NOT_FOUND;
+      }
+
+      if (result != null) {
+        if (fields == null){
+          while (resultSet.next()){
+            String field = resultSet.getString(2);
+            String value = resultSet.getString(3);
+            result.put(field, new StringByteIterator(value));
+          }
+        } else {
+          for (String field : fields) {
+            String value = resultSet.getString(field);
+            result.put(field, new StringByteIterator(value));
+          }
+        }
+      }
+      resultSet.close();
+      return Status.OK;
+
+    } catch (SQLException e) {
+      LOG.error("Error in processing read of table " + tableName + ": " + e);
+      return Status.ERROR;
+    }
+  }
+
+  @Override
+  public Status scan(String tableName, String startKey, int recordcount, Set<String> fields,
+                     Vector<HashMap<String, ByteIterator>> result) {
+    try {
+      PreparedStatement scanStatement = connection.prepareStatement(createScanStatement(tableName, fields));
+
+      scanStatement.setString(1, startKey);
+      scanStatement.setInt(2, recordcount);
+      ResultSet resultSet = scanStatement.executeQuery();
+
+      for (int i = 0; i < recordcount && resultSet.next(); i++) {
+        if (result != null && fields != null) {
+          HashMap<String, ByteIterator> values = new HashMap<String, ByteIterator>();
+          for (String field : fields) {
+            String value = resultSet.getString(field);
+            values.put(field, new StringByteIterator(value));
+          }
+
+          result.add(values);
+        }
+      }
+
+      resultSet.close();
+      return Status.OK;
+    } catch (SQLException e) {
+      LOG.error("Error in processing scan of table: " + tableName + ": " + e);
+      return Status.ERROR;
+    }
+  }
+
+  @Override
+  public Status update(String tableName, String key, Map<String, ByteIterator> values) {
+    try{
+      JSONObject jsonObject = new JSONObject();
+      for (Map.Entry<String, ByteIterator> entry : values.entrySet()) {
+        jsonObject.put(entry.getKey(), entry.getValue().toString());
+      }
+
+      PreparedStatement updateStatement = connection.prepareStatement(createUpdateStatement(tableName));
```

Comment:
Shouldn't these prepared statements be cached? Plenty of other implementations have left this kind of optimization for later, I'm just curious if there's some special consideration in building them each time for now.